### PR TITLE
add flower painting tutorial

### PR DESCRIPTION
**Pull request by andrewhinh**
Adds a flower painting tutorial inspired by [this tweet](https://x.com/kickingkeys/status/2091570990048276897) and subsequent inspired [blog post](https://huggingface.co/blog/train-to-paint-with-code) that replaces the [existing sandboxes tutorial](https://gym.modal.dev/tutorials/sandboxes).

Basically a cleaned up version of https://github.com/modal-projects/training-gym/pull/473.

Example of a [successful run](https://modal-labs-ajhinh-dev--training-gym-dashboard-fastapi-app.modal.run/training/shy-tuple-ca0e7a013267?tab=summary).

---

<a href="https://modal.devinenterprise.com/review/modal-projects/training-gym/pull/538" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-devin-review-dark.svg?v=3">
    Devin Review
  </picture>
</a>

### CODE REVIEW

File: tutorials/paint_flowers/build_pool.py [hunk 1]
```diff
@@ -0,0 +1,629 @@
...
+        )
+        refs[f"{name.split('_', 1)[1]}.webp"] = buf.getvalue()
+    return refs, sorted(chosen)
+
+
+@app.function(image=POOL_IMAGE, timeout=7200, cpu=4.0, memory=8192, serialized=True)
+def build_pool_remote(corpus: str = "/tmp/flower-corpus") -> dict:
+    dest = pathlib.Path(corpus)
+    if not any(dest.glob("pos*.png")):
+        render_corpus(dest)
```
[the remark this answers is not in the document]
Comment (andrewhinh) [reply tutorials/paint_flowers/build_pool.py]: fixed in e92c61f8904e414ce60416877a8e5b77e241a263

File: tutorials/paint_flowers/build_pool.py [hunk 2]
```diff
@@ -0,0 +1,629 @@
...
+                (
+                    f"pos{len(items):03d}_{species}_{palette}_{seed}",
+                    flower_sketch(species, palette, seed=seed),
+                )
+            )
+    items += make_negatives(84, seed=3)
+    out.mkdir(parents=True, exist_ok=True)
+    render_app = modal.App.lookup(RENDER_APP_NAME, create_if_missing=True)
+    with cf.ThreadPoolExecutor(sandboxes) as created:
+        sbs = list(created.map(lambda _: open_sandbox(render_app), range(sandboxes)))
```
[the remark this answers is not in the document]
Comment (andrewhinh) [reply tutorials/paint_flowers/build_pool.py]: fixed in 672d2a28f1f7c5cba5da2cf1142b9dd9da50e441

File: tutorials/paint_flowers/helpers.py [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+      ? window.color(v) : v;
+    return realLerp(c(a), c(b), t);
+  };
+  for (const name of ["pick", "spline", "flowLine", "polygon", "hatch",
+                      "noHatch", "setHatch", "bleed", "field", "noField"]) {
+    if (window[name] === undefined) window[name] = (...a) => facade[name](...a);
+  }
+})();
+</script>
+<script>try{${sketch}}catch(e){window.__err=String(e)}</script>
```
[the remark this answers is not in the document]
Comment (andrewhinh) [reply tutorials/paint_flowers/helpers.py]: fixed in 672d2a28f1f7c5cba5da2cf1142b9dd9da50e441

File: tutorials/paint_flowers/helpers.py [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+    parsed = parse_response(response)
+    m = _JS_FENCE.search(parsed.content or "")
+    if not m:
+        return None
+    code = m.group(1).strip()
+    if not code or "function setup" not in code or len(code) > 8000:
+        return None
+    if _BANNED.search(code):
+        return None
+    return code
```
[the remark this answers is not in the document]
Comment (andrewhinh) [reply tutorials/paint_flowers/helpers.py:30]: fixed in 672d2a28f1f7c5cba5da2cf1142b9dd9da50e441

File: tutorials/paint_flowers/helpers.py [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+      ? window.color(v) : v;
+    return realLerp(c(a), c(b), t);
+  };
+  for (const name of ["pick", "spline", "flowLine", "polygon", "hatch",
+                      "noHatch", "setHatch", "bleed", "field", "noField"]) {
+    if (window[name] === undefined) window[name] = (...a) => facade[name](...a);
+  }
+})();
+</script>
+<script>try{(0,eval)(${JSON.stringify(sketch).replace(/</g, '\\u003c')})}catch(e){window.__err=String(e)}</script>
```
[the remark this answers is not in the document]
Comment (andrewhinh) [reply tutorials/paint_flowers/helpers.py]: fixed in e7d906b4711c7ad500d79dc4c4ab2b1e8200230c

File: tutorials/paint_flowers/main.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+# as the reference pool. During each rollout, sketches are rendered to PNGs in a
+# [Modal Sandbox](https://modal.com/docs/guide/sandboxes) and
+# [Qwen3.6-27B](https://huggingface.co/Qwen/Qwen3.6-27B) compares each against the
+# reference image pool.
+
+import asyncio
+import itertools
+import random
+
+import helpers
```
[the remark this answers is not in the document]
Comment (andrewhinh) [reply tutorials/paint_flowers/main.py:29]: fixed in 259d74a1488acbd4e71469f433305ab9eef76e0c

File: tutorials/paint_flowers/main.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+# as the reference pool. During each rollout, sketches are rendered to PNGs in a
+# [Modal Sandbox](https://modal.com/docs/guide/sandboxes) and
+# [Qwen3.6-27B](https://huggingface.co/Qwen/Qwen3.6-27B) compares each against the
+# reference image pool.
+
+import asyncio
+import itertools
+import random
+
+from . import helpers
```
[the remark this answers is not in the document]
Comment (andrewhinh) [reply tutorials/paint_flowers/main.py]: fixed in 9a3c29e814cfdda15b4e5953e63cbdc7a0c1ffc2

File: tutorials/paint_flowers/helpers.py [hunk 8]
```diff
@@ -0,0 +1,523 @@
...
+    }
+  };
+
+  try {
+    const res = await attempt(sketch);
+    if (!res.buf) {
+      console.error('SKETCH_ERROR: ' + (res.err || 'no canvas'));
+      process.exit(2);
+    }
+    process.stdout.write('PNGB64:' + Buffer.from(res.buf).toString('base64'));
```
[the remark this answers is not in the document]
Comment (andrewhinh) [reply tutorials/paint_flowers/helpers.py:148]: intentional

**Comment by joyliu-q** [comment]
I dont see the diff between reward 0 on final run here vs reward 90% https://modal-labs-ajhinh-dev--training-gym-dashboard-fastapi-app.modal.run/training/shy-tuple-ca0e7a013267?tab=rollouts

File: tutorials/sandboxes.py [hunk 9]
```diff
@@ -1,198 +0,0 @@
...
-import modal
-
-from modal_training_gym import (
-    Endpoint,
-    HarborDataset,
-    Qwen3_5_4B,
-    Qwen3_5_4B_Recipe,
-    TrainConfig,
-    extract_code,
-)
```
Comment (joyliu-q) [inline tutorials/sandboxes.py:30]: Why did we delete this? Driveby?

Comment (andrewhinh) [reply tutorials/sandboxes.py:30]: no, the flower tutorial is meant to replace this one. the sandboxes tutorial is a bit jank and overlaps in concept with https://github.com/modal-projects/training-gym/pull/505. also, this tutorial covers the use of Modal Sandboxes.

**Comment by andrewhinh** [comment]
> I dont see the diff between reward 0 on final run here vs reward 90% [modal-labs-ajhinh-dev--training-gym-dashboard-fastapi-app.modal.run/training/shy-tuple-ca0e7a013267?tab=rollouts](https://modal-labs-ajhinh-dev--training-gym-dashboard-fastapi-app.modal.run/training/shy-tuple-ca0e7a013267?tab=rollouts)

responded in slack

**Review by joyliu-q — APPROVED** [review]

File: tutorials/paint_flowers/main.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+            for r in self.prompts
+        ]
+
+
+N_TRAIN = 224
+N_EVAL = 8
+combos = list(itertools.product(SPECIES, PALETTES))
+random.Random(7).shuffle(combos)
+train_dataset = FlowerPromptDataset(build_prompts(combos, N_TRAIN))
+eval_dataset = FlowerPromptDataset(build_prompts(combos, N_EVAL))
```
[the remark this answers is not in the document]
Comment (andrewhinh) [reply tutorials/paint_flowers/main.py:154]: is fine, the point is we can visually look at the flowers and see them getting better (i.e., look at your data!)
